### PR TITLE
build!: sanitizers for gcc

### PR DESCRIPTION
**Pull request by ii14**
GCC also supports sanitizers. GCC doesn't support -fsanitize-blacklist option though, so replace .asan-blacklist file with no_sanitize_address function attributes instead.

### CODE REVIEW

**Comment by dundargoc** [comment]
Adress sanitizer works on MSVC as well, but I can take care of that later on.

File: CMakeLists.txt [hunk 1]
```diff
@@ -146,8 +146,10 @@ if((CLANG_ASAN_UBSAN AND CLANG_MSAN)
   message(FATAL_ERROR "Sanitizers cannot be enabled simultaneously")
 endif()
 
-if((CLANG_ASAN_UBSAN OR CLANG_MSAN OR CLANG_TSAN) AND NOT CMAKE_C_COMPILER_ID MATCHES "Clang")
-  message(FATAL_ERROR "Sanitizers are only supported for Clang")
+if(CLANG_ASAN_UBSAN OR CLANG_MSAN OR CLANG_TSAN)
```
[the remark this answers is not in the document]
Comment (ii14) [reply CMakeLists.txt:149]: I'm all for it.

**Review by justinmk — APPROVED** [review]

**Comment by justinmk** [comment]
Do we need a test run with e.g. an intentional leak/overrun, to make sure CI still correctly catches stuff?

**Comment by ii14** [comment]
@justinmk Just tested it, and ASAN still catches leaks on CI

**Comment by zeertzjq** [comment]
This is a breaking change. Should this be documented somewhere?

**Comment by ii14** [comment]
Now that the "following HEAD" issue is gone, is news.txt a right place for this?

Maybe setting `CLANG_*` options could just fail the build?

**Comment by dundargoc** [comment]
> Now that the "following HEAD" issue is gone, is news.txt a right place for this?

Nah, no users care about this

> Maybe setting CLANG_* options could just fail the build?

Too complicated. Don't bother IMO.

**Comment by dundargoc** [comment]
> This is a breaking change. 

Breaking for whom? Neovim developers will either expect this, or readjust.

> Should this be annouced somewhere?

Nah.

**Comment by zeertzjq** [comment]
> Nah, no users care about this

Why?

**Comment by dundargoc** [comment]
@ii14 if you think users will be interested to know about this then sure, add an entry to news.txt. Calling @clason to chime in as he's our newsanchor.

I am hard-vetoing `CLANG_*` options failing the build. We should not add long-term solutions to short-term problems.

**Comment by bfredl** [comment]
What is considered to be the breaking change here? just the name of cmake VARS being changed?

**Comment by zeertzjq** [comment]
> What is considered to be the breaking change here? just the name of cmake VARS being changed?

Yes.

**Comment by dundargoc** [comment]
I interpreted the breaking change label/convention to be for user-breaking changes, not developer-breaking changes. Even if we don't write a news entry, the changelog at releases would be filled with "changed variable names" in the "breaking changes" which I can't imagine any user having interest in.

**Comment by justinmk** [comment]
Changes that affect "users" that build from source should usually be noted in news.txt. I don't see any harm in that, and it's useful for package maintainers, which are also "users".

**Comment by dundargoc** [comment]
> Changes that affect "users" that build from source should usually be noted in news.txt. I don't see any harm in that, and it's useful for package maintainers, which are also "users".

The harm would be that news.txt becomes a glorified changelog, with extraneous information most people won't care about which would defeat its purpose.

If this is the route you want to go with, then we should probably retroactively add all breaking build system changes to news.txt to be consistent? I've changed A LOT of variables between 0.8 and master.

**Comment by justinmk** [comment]
> If this is the route you want to go with, then we should probably retroactively add all breaking build system changes to news.txt to be consistent? I've changed A LOT of variables between 0.8 and master.

No. It's a judgement call depending on the prominence and importance of the change. Since it was called out here, presumably it's more important than past changes. We don't need to revisit and re-litigate user vs low-level. If we don't think this change is worth calling out in news.txt, then don't. It's not a blocker, as I mentioned ("no strong opinion").

**Pull request merged**